sky/provision/provisioner.py

```diff
@@ -360,25 +365,43 @@ def wait_for_ssh(cluster_info: provision_common.ClusterInfo,
     ip_list = cluster_info.get_feasible_ips()
     port_list = cluster_info.get_ssh_ports()
 
+
+
+    while True:
+        task = queue.get()
+        if task is None:
+            break
+
     timeout = 60 * 10  # 10-min maximum timeout
     start = time.time()
-    # use a queue for SSH querying
-    ips = collections.deque(ip_list)
-    ssh_ports = collections.deque(port_list)
-    while ips:
-        ip = ips.popleft()
-        ssh_port = ssh_ports.popleft()
-        success, stderr = waiter(ip, ssh_port, **ssh_credentials)
-        if not success:
-            ips.append(ip)
-            ssh_ports.append(ssh_port)
+    queue = queue.Queue()
+    
+    def _worker_thread(queue):
+        while True:
+            task = queue.get()
+            if task is None:
+                break
+            ip, ssh_port = task
+            success, stderr = waiter(ip, ssh_port, **ssh_credentials)
             if time.time() - start > timeout:
                 with ux_utils.print_exception_no_traceback():
                     raise RuntimeError(
                         f'Failed to SSH to {ip} after timeout {timeout}s, with '
                         f'{stderr}')
             logger.debug('Retrying in 1 second...')
             time.sleep(1)
+    
+    with concurrent.futures.ThreadPoolExecutor(max_workers=constants.NUM_THREADS) as executor:
+        for _ in range(constants.NUM_THREADS):
+            executor.submit(_worker_thread, queue)
+
+        for ip, ssh_port in zip(ips, ssh_ports):
+            queue.put((ip, ssh_port))
+
+        queue.join()
```

Comment:
We actually use `subprocess_utils.run_in_parallel` in most of codepath. Maybe we should use that here as well.
https://github.com/skypilot-org/skypilot/blob/3c3bcee5cfe720a96ab67f4049a557a79e7f077f/sky/backends/backend_utils.py#L1346

Reply:
I was thinking of reducing the resource usage, maybe it would be better if we start the multiprocessing waiter only after one node succeeded.